### PR TITLE
Add generated catalogue models

### PR DESCRIPTION
**Pull request by michpohl**
This adds generated catalog**ue** code made using [openapi-generator-cli](https://openapi-generator.tech/docs/installation/)from our catalog API endpoint definitions.
Only the generated models have been added so far. Generated helper class have issues with maitaining compatibility with Android SDKs <26, so I left them out for now.

This is a temporary version intended for prototyping. It will be subject to change.
The generated code has a few small issues, but if you're looking to use just the types, you'll be fine.

This also adds `catalogue-openapi.json`, our current state of the ASPI definition,
and `openapi-config.yml`, specifying the generator-specific args used for generation.
This was the actual command used, from SDK root:
```
openapi-generator generate -i catalogue/catalogue_openapi.json -g kotlin -o ./catalogue/ -c catalogue/openapi-config/openapi-config.yml
```

This will also not be released in a public package yet - some improvements need to be made for a first "official" release. 
If you have trouble creating a local release to work with this hit me up, and I'll show you how.

### CODE REVIEW

File: catalog/build.gradle.kts [hunk 1]
```diff
@@ -10,12 +10,17 @@ android {
 
 dependencies {
 
+    implementation(libs.kotlinxCoroutinesCore)
+    implementation(libs.kotlin.reflect)
```
Comment (CanYumusak) [inline catalog/build.gradle.kts:14]: Do we really need reflection? :/

File: catalog/proguard-rules.pro [hunk 2]
```diff
@@ -0,0 +1,11 @@
+-keepattributes *Annotation*, InnerClasses
+-dontnote kotlinx.serialization.AnnotationsKt # core serialization annotations
+
```
Comment (CanYumusak) [inline catalog/proguard-rules.pro:3]: why do we need these rules?

File: catalog/proguard-rules.pro [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+# kotlinx-serialization-json specific. Add this if you have java.lang.NoClassDefFoundError kotlinx.serialization.json.JsonObjectSerializer
+-keepclassmembers class kotlinx.serialization.json.** { *** Companion; }
+-keepclasseswithmembers class kotlinx.serialization.json.** { kotlinx.serialization.KSerializer serializer(...); }
+
+# project specific.
+-keep,includedescriptorclasses class org.openapitools.client.models.**$$serializer { *; }
+-keepclassmembers class org.openapitools.client.models.** { *** Companion; }
```
Comment (CanYumusak) [inline catalog/proguard-rules.pro:10]: why do we need these rules?

File: catalog/src/main/kotlin/org/openapitools/client/apis/AlbumJSONAPI.kt [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+ */
+
+@file:Suppress(
+    "ArrayInDataClass",
+    "EnumEntryName",
+    "RemoveRedundantQualifierName",
+    "UnusedImport"
+)
+
+package org.openapitools.client.apis
```
Comment (CanYumusak) [inline catalog/src/main/kotlin/org/openapitools/client/apis/AlbumJSONAPI.kt:16]: Should we use our packagename instead?

File: catalog/src/main/kotlin/org/openapitools/client/infrastructure/ApiClient.kt [hunk 5]
```diff
@@ -0,0 +1,277 @@
+package org.openapitools.client.infrastructure
+
+import okhttp3.OkHttpClient
```
Comment (CanYumusak) [inline catalog/src/main/kotlin/org/openapitools/client/infrastructure/ApiClient.kt:3]: Consider using Retrofit instead?

File: catalog/src/main/kotlin/org/openapitools/client/infrastructure/ApiClient.kt [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+        protected const val Authorization = "Authorization"
+        protected const val JsonMediaType = "application/json"
+        protected const val FormDataMediaType = "multipart/form-data"
+        protected const val FormUrlEncMediaType = "application/x-www-form-urlencoded"
+        protected const val XmlMediaType = "application/xml"
+        protected const val OctetMediaType = "application/octet-stream"
+
+        val apiKey: MutableMap<String, String> = mutableMapOf()
+        val apiKeyPrefix: MutableMap<String, String> = mutableMapOf()
+        var username: String? = null
```
Comment (CanYumusak) [inline catalog/src/main/kotlin/org/openapitools/client/infrastructure/ApiClient.kt:50]: any way to remove basic auth from the config?

**Review by CanYumusak — APPROVED** [review]

File: catalog/build.gradle.kts (hunk 1 above)
Comment (michpohl) [reply catalog/build.gradle.kts:14]: Nope, omission on my end. Thanks for catching!

File: catalog/proguard-rules.pro (hunk 2 above)
Comment (michpohl) [reply catalog/proguard-rules.pro:3]: They are generated. At this point, I want to change as little as possible in the generated files.

File: catalog/src/main/kotlin/org/openapitools/client/apis/AlbumJSONAPI.kt (hunk 4 above)
Comment (michpohl) [reply catalog/src/main/kotlin/org/openapitools/client/apis/AlbumJSONAPI.kt:16]: Yes. I couldn't convince the tool to do that without issues though. So for this first version, I want to use theirs. 

I intend to update this later, however.

File: catalog/src/main/kotlin/org/openapitools/client/infrastructure/ApiClient.kt (hunk 5 above)
Comment (michpohl) [reply catalog/src/main/kotlin/org/openapitools/client/infrastructure/ApiClient.kt:3]: If you look at the commit message, I have considered and instructed the generator to do to so 🙂

However, I do not intend to keep the API client as is anyways, as it doesn't fit in with our SDK ecosystem.

File: catalog/proguard-rules.pro (hunk 3 above)
Comment (michpohl) [reply catalog/proguard-rules.pro:10]: Same reason as the other rules 🙂

**Comment by stefanmoro** [comment]
Rename to catalogue please :)

**Comment by michpohl** [comment]
> Rename to catalogue please :) We follow the same naming as the corresponding API domain: https://developer.tidal.com/reference/web-api?spec=catalogue-v2&ref=get-albums-v2

Will do! This info came after the modules had been created and I will fix the naming in a separate PR for both iOS and Android! Good?

File: catalog/build.gradle.kts (hunk 1 above)
Comment (michpohl) [reply catalog/build.gradle.kts]: removed

**Review by CanYumusak — APPROVED** [review]

**Comment by michpohl** [comment]
Ok, so I'm not merging this fir now. The reason:
`openapi-generator` doesn't honor the compatibility to API 25 and below, even when the setting is on. Looks like the developers forgot a few classes to cover here...

As you can see above, lint points this out - and initially I wanted to just override this, but since no one uses this yet, I will first do another look at what I can do about it.

**Comment by michpohl** [comment]
@CanYumusak I found issues with the generator and therefore opted to only push the models for now (as is requested)

**Review by CanYumusak — APPROVED** [review]

**Pull request merged**